Add baseline enforcement within the runtime

vlovich opened the pull request:
Middleware might restrict the size of the URL that would get generate. Set a sane unrealistic upper bound that's guaranteed to get through to the KV worker. This should make error messages reliably clear when a name provided is too long.

bretthoerner commented on src/workerd/api/kv.c++:
```diff
@@ -26,6 +26,8 @@ static void validateKeyName(kj::StringPtr name) {
   JSG_REQUIRE(name != ".", TypeError, "\".\" is not allowed as a key name.");
   JSG_REQUIRE(name != "..", TypeError, "\"..\" is not allowed as a key name.");
+  JSG_REQUIRE(name.size() <= 8 * 1024, TypeError, "UTF-8 encoded length of ", name.size(),
+      " exceeds limit of 8 KiB.");
```
I'd go ahead and add a mention to "key name" here, since it's a little ambiguous otherwise.

And are we sure we want a higher limit here? It's a little confusing to get an error about `8 KiB`, then you try a 7 KiB key and get a totally different error, with a different key length. We haven't changed KV key length in forever (and I can't remember a single request about it), any reason not to just duplicate the same length/error here and worry about the tiny chance it changes later?

vlovich replied:
Done.

bretthoerner replied:
Any thoughts on the second bit?

jclee replied:
(possible that the error message expression may need to be `kj::str("UTF-8 encoded length of ", name.size(), " exceeds limit of 8 KiB.")` to ensure that it doesn't attempt to stringify `name.size() = ` into the message?)

vlovich replied:
Nope. `JSG_REQUIRE` macros are different & don't do the normal `KJ` stringification.

koeninger commented:
Isn't this a breaking change unless it's literally exactly the same error text that's currently thrown by the KV worker for keys in the  512b - 32k range?

vlovich commented:
I had a concern that the message would be different (hence why I initially chose 8 kib - the probability of you seeing it is minimized unless you're really going crazy) but what's not clear to me is how it's a breaking change. If this is a concern I'd rather go back to 8 kib or put a 32 kib limit here rather than trying to massage the error message to be identical as that seems harder to accomplish / harder to test for.

Thinking about this some more, the bigger breaking change concern I have here is that this:

```
NAMESPACE.get(<very long string>)
```
will now throw whereas before you'd need to await because it was coming from the KV worker:
```
NAMESPACE.get(<very long string>).catch(...)
```

I can add a separate `js.evalNow` to wrap the entrypoint functions because it looks like the KV runtime hadn't been doing that whereas @kentonv had indicated all entrypoint methods returning a promise should be doing that. Not sure the consequence of that as it will subtly change some of the exception generation but historically IIRC we've just gone ahead and done that without a compat flag.

jasnell commented on src/workerd/api/kv.c++:
```diff
@@ -26,6 +26,8 @@ static void validateKeyName(kj::StringPtr name) {
   JSG_REQUIRE(name != "", TypeError, "Key name cannot be empty.");
   JSG_REQUIRE(name != ".", TypeError, "\".\" is not allowed as a key name.");
   JSG_REQUIRE(name != "..", TypeError, "\"..\" is not allowed as a key name.");
+  JSG_REQUIRE(name.size() <= 512, TypeError, "UTF-8 encoded length of ", name.size(),
```
Nit: For hard coded limits like this, I'd much prefer a constant defined. Non-blocking tho.

More importantly, some comment here as to why the limit is 512 would be helpful for future generations.

bretthoerner commented:
> What's not clear to me is how it's a breaking change - an error is an error and we have never bothered with error message stability as a back compat guarantee.

Dunno what to say, we've just been careful about it in KV. For better or worse the message includes an HTTP status code and there is definitely code in the wild that string matches on them. I do agree that it'd be pretty weird for someone to catch long key lengths, when they could easily check that in JS before calling KV methods, but :shrug: 

But yeah, throwing synchronously and not after an `await` feels like a bigger change.

vlovich commented on src/workerd/api/kv.c++:
```diff
@@ -26,6 +26,8 @@ static void validateKeyName(kj::StringPtr name) {
   JSG_REQUIRE(name != "", TypeError, "Key name cannot be empty.");
   JSG_REQUIRE(name != ".", TypeError, "\".\" is not allowed as a key name.");
   JSG_REQUIRE(name != "..", TypeError, "\"..\" is not allowed as a key name.");
+  JSG_REQUIRE(name.size() <= 512, TypeError, "UTF-8 encoded length of ", name.size(),
```
See the above comment. `JSG_REQUIRE` doesn't format like `KJ` does.

```
#define JSG_REQUIRE(cond, jsErrorType, ...)                                             \
  KJ_REQUIRE(cond, kj::str(JSG_EXCEPTION(jsErrorType) ": ", ##__VA_ARGS__))
```

vlovich commented:
I think I've addressed the comments.

1. Extracted out constant to a standalone variable.
2. Mimic the error response to be similar to what it would look like if FL rejected it.
3. Wrap the code in `js::evalNow` to fix the preexisting issue that exceptions were thrown at the call site instead of within the promise.

koeninger commented:
> similar to what it would look like if FL rejected it.

It's not FL rejecting it that's the common case for 512b < size < 32k, so it's sgw's error message that we care about being the same as, right?

bretthoerner commented on src/workerd/api/kv.c++:
```diff
@@ -22,10 +25,13 @@ static void checkForErrorStatus(kj::StringPtr method, const kj::HttpClient::Resp
   JSG_REQUIRE(name != "..", TypeError, "\"..\" is not allowed as a key name.");
+  JSG_REQUIRE(name.size() <= kMaxKeyLength, Error, "KV ", method, " failed: ",
+      414, " Request-URI Too Large. UTF-8 encoded key length of ", name.size(),
+      " exceeds the 512 byte limit.");
```
Any reason not to mirror what SGW would return?

```
KV GET failed: 414 UTF-8 encoded length of 13216 exceeds key length limit of 512.
```

Also may as well use `kMaxKeyLength` in place of `512`

bretthoerner reviewed (approved):
I think you may as well match up the error since it's easy, but LGTM.

kentonv commented:
FYI, there is actually a compat change which causes JSG to convert synchronous exceptions into async exceptions across _all_ APIs that return promises, which became default 2022-10-31. So the introduction of `evalNow()` in this PR is a noop for anyone who has that flag set.
